disperser/controller/controller.go

```diff
@@ -466,36 +480,28 @@ func (c *Controller) logAttestationUpdate(batchHeaderHash string, quorumResults
 		"quorumPercentages", quorumPercentagesBuilder.String())
 }
 
-// Iterates over the input metadata slice, and returns a new slice with stale and duplicate metadatas filtered out
-func (c *Controller) filterStaleAndDedupBlobs(
-	ctx context.Context,
-	inputMetadatas []*v2.BlobMetadata,
-) []*v2.BlobMetadata {
-	outputMetadatas := make([]*v2.BlobMetadata, 0, len(inputMetadatas))
-	now := c.getNow()
-
-	for _, metadata := range inputMetadatas {
-		blobKey, err := metadata.BlobHeader.BlobKey()
-		if err != nil {
-			c.logger.Errorf("compute blob key: %w", err)
-			// we must discard if we cannot compute key, since it's used for deduplication
-			continue
-		}
-
-		if c.checkAndHandleStaleBlob(ctx, blobKey, now, metadata.BlobHeader.PaymentMetadata.Timestamp) {
-			// discard stale blob
-			continue
-		}
+// Returns true if the blob is both unique (i.e. not already being dispatched) and fresh (i.e. not stale).
+// This method has the side effect of adding the blob to the blobSet, and updating metrics.
+func (c *Controller) isUniqueAndFresh(ctx context.Context, blobMetadata *v2.BlobMetadata) bool {
+	blobKey, err := blobMetadata.BlobHeader.BlobKey()
+	if err != nil {
+		c.logger.Errorf("compute blob key: %w", err)
+		// we must discard if we cannot compute key, since it's used for deduplication
+		return false
+	}
 
-		if c.blobSet.Contains(blobKey) {
-			// discard duplicate blob
-			continue
-		}
+	if c.checkAndHandleStaleBlob(ctx, blobKey, c.getNow(), blobMetadata.BlobHeader.PaymentMetadata.Timestamp) {
+		// discard stale blob
+		return false
+	}
 
-		outputMetadatas = append(outputMetadatas, metadata)
+	if c.blobSet.Contains(blobKey) {
+		// discard duplicate blob
+		return false
 	}
+	c.blobSet.AddBlob(blobKey)
```

Comment:
You're adding to the `blobSet` here, but there are cases within `NewBatch` where something can fail before the blob is _actually_ put into a batch.

That means that the duplicate check would fail when the blob is retried. I think in such a case, the blob would forever be in the `blobSet`, and never get cleaned up.

I think you might need to delay adding the blob to the blob set until after the blob has been batched.

Somehow the _180_ line `NewBatch` method is tricky to fully comprehend...